api_common.go

```diff
@@ -3,9 +3,76 @@ package iotago
 import (
 	"context"
 
+	"github.com/iotaledger/hive.go/ierrors"
+	"github.com/iotaledger/hive.go/serializer/v2"
 	"github.com/iotaledger/hive.go/serializer/v2/serix"
 )
 
+var (
+	// the addresses need to be unique and lexically ordered to calculate a deterministic bech32 address for a MultiAddress.
+	// HINT: the uniqueness is checked within a custom validator function, which is on MultiAddress level.
+	addressesWithWeightV3ArrRules = &serix.ArrayRules{
+		Min:            1,
+		Max:            10,
+		ValidationMode: serializer.ArrayValidationModeLexicalOrdering,
+	}
+
+	// multiAddressValidatorFunc is a validator which checks that:
+	//  1. MultiAddresses are not nested inside the MultiAddress.
+	//  2. "raw address part" of all addresses are unique (without type byte and capabilities).
+	//  3. The weight of each address is at least 1.
+	//  4. The threshold is smaller or equal to the cumulative weight of all addresses.
+	multiAddressValidatorFunc = func(ctx context.Context, addr MultiAddress) error {
+		addrSet := map[string]int{}
+
+		var cumulativeWeight uint16
+		for idx, address := range addr.Addresses {
+			var addrWithoutTypeAndCapabilities []byte
+			switch addr := address.Address.(type) {
+			case *AccountAddress:
+				addrWithoutTypeAndCapabilities = addr[:]
+			case *Ed25519Address:
+				addrWithoutTypeAndCapabilities = addr[:]
+			case *ImplicitAccountCreationAddress:
+				addrWithoutTypeAndCapabilities = addr[:]
+			case *MultiAddress:
+				return ierrors.Wrapf(ErrNestedMultiAddress, "address with index %d is a multi address inside a multi address", idx)
+			case *NFTAddress:
+				addrWithoutTypeAndCapabilities = addr[:]
+			case *RestrictedEd25519Address:
+				addrWithoutTypeAndCapabilities = addr.PubKeyHash[:]
+			default:
+				return ierrors.Wrapf(ErrUnknownAddrType, "address with index %d has an unknown address type (%T) inside a multi address", idx, addr)
+			}
+
+			// we need to check for uniqueness of the address, but instead of the whole serialized address, we need to ignore
+			// different address types or capabilities, that might result in the same signature.
+			addrString := string(addrWithoutTypeAndCapabilities)
+			if j, has := addrSet[addrString]; has {
+				return ierrors.Wrapf(serializer.ErrArrayValidationViolatesUniqueness, "element %d and %d are duplicates", j, idx)
```

Comment:
```suggestion
				return ierrors.Wrapf(serializer.ErrArrayValidationViolatesUniqueness, "addresses with indices %d and %d in multi address are duplicates", j, idx)
```

Nit: More helpful/precise error message.